### PR TITLE
Usage rules in crac api

### PR DESCRIPTION
**Pull request by baptsegu**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Three big changes : 
- put in crac-api sub-interfaces of the UsageRule interface : FreeToUse and OnState
- change the logic of the FreeToUse UsageRule, which now contains an Instant instead of a State
- delete the OnContingency UsageRule : not used for now

**Does this PR introduce a breaking change?** *(What changes might users need to make in their application due to this PR?)*

Yes, breaking change in how FreeToUse UsageRules are defined

### CODE REVIEW

File: ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/RaoInputHelper.java [hunk 1]
```diff
@@ -116,6 +116,8 @@ public static void synchronize(Crac crac, Network network) {
         });
         noValidAction.forEach(networkAction -> crac.removeNetworkAction(networkAction.getId()));
 
+        // remove onContingency Remedial Action
```
Comment (phiedw) [inline ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/RaoInputHelper.java:119]: is this a TODO?

**Review by phiedw — CHANGES_REQUESTED** [review]

File: ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/RaoInputHelper.java (hunk 1 above)
Comment (baptsegu) [reply ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/RaoInputHelper.java:119]: yes :), this is the initial purpose of the PR, but in the end I decided to divide it in another PR for them to be easier to read. You are right i will put a TODO its not clear

File: data/crac/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/SimpleCrac.java [hunk 2]
```diff
@@ -430,13 +431,11 @@ public PstRangeActionAdder newPstRangeAction() {
     }
 
     public void addNetworkAction(NetworkAction networkAction) {
-        networkAction.getUsageRules().forEach(usageRule -> addState(usageRule.getState()));
```
[the remark this answers is not in the document]
Comment (baptsegu) [reply data/crac/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/SimpleCrac.java:433]: Indeed, I wanted to do something clean with an Adder pattern on the usageRule, but postponed it as it is not a priority. But you are right I should do something here. I will add it and adapt the unit-test

[the remark this answers is not in the document]
Comment (baptsegu) [reply data/crac/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/SimpleCrac.java:433]: It's actually no straightforward as the SimpleCrac adders really lack consistency. For instance : 
* `the addNetworkAction()` method adds the states of the RA's usage rules and its network elements in the SimpleCrac 
* `the addRangeAction()` method does not that. It only adds the RA without checking that the subelements are in the SimpleCrac.

And : 
* the `addInstant(Instant newInstant)` returns an expection if newInstant is already in the SimpleCrac
* the `addState`, or `addCnec` does not return an exception in that case, and just don't create a duplicated element in the SimpleCrac

And : 
* Some adders add the subelements of an object in the SimpleCrac. For instance addNetworkAction adds the states of the UR of the network action.
* Some methods does not that, for instance addState returns an exception if the instant or the contingency of the States does not exists.

Even funnier : the different addCnec() methods have different behaviors, some adds the NetworkElement in the SimpleCrac, some other returns exception if the SimpleCrac does not contains the NetworkElement already.


So it's really hard to know which behaviour should be expected here.
I propose to keep it like this (we are actually doing the same thing as RangeAction), and wait for a refactoring of crac-impl to align the behaviour of all the adders.

File: data/crac/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/usage_rule/FreeToUseImpl.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import com.farao_community.farao.data.crac_api.usage_rule.UsageMethod;
+import com.farao_community.farao.data.crac_impl.json.serializers.usage_rule.FreeToUseSerializer;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+
+/**
+ * The UsageMethod of the FreeToUseImpl UsageRule is effective in all the States which
+ * are at a given Instant.
+ *
+ * @author Viktor Terrier {@literal <viktor.terrier at rte-france.com>}
```
[the remark this answers is not in the document]
Comment (baptsegu) [reply data/crac/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/usage_rule/FreeToUseImpl.java:22]: ok ok, I added my name

File: data/crac/crac-impl/src/test/java/com/farao_community/farao/data/crac_impl/CracFileTest.java [hunk 4]
```diff
@@ -574,11 +573,9 @@ public void testAddRangeActionWithNoConflict() {
         Mockito.when(instant.getId()).thenReturn("instantid");
         Mockito.when(state.getInstant()).thenReturn(instant);
         Mockito.when(state.getContingency()).thenReturn(Optional.empty());
-        Mockito.when(rangeAction.getUsageRules()).thenReturn(Collections.singletonList(new FreeToUse(UsageMethod.AVAILABLE, state)));
 
         simpleCrac.addRangeAction(rangeAction);
 
-        assertNotNull(simpleCrac.getPreventiveState());
```
[the remark this answers is not in the document]
Comment (baptsegu) [reply data/crac/crac-impl/src/test/java/com/farao_community/farao/data/crac_impl/CracFileTest.java:581]: cf. answer above. Not sure what to do here as there is no clear pattern in the SimpleCrac

**Comment by baptsegu** [comment]
Thanks for the review @phiedw , i made some changes

**Review by phiedw — APPROVED** [review]

**Pull request merged**